File: AetherPlayerShop/src/main/java/loyfael/commands/ChestShopCommand.java
package loyfael.commands;

import loyfael.Main;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Commande principale du plugin AetherPlayerShop
 */
public class ChestShopCommand implements CommandExecutor, TabCompleter {

    private final Main plugin;

    public ChestShopCommand(Main plugin) {
        this.plugin = plugin;
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command,
                           @NotNull String label, @NotNull String[] args) {

        if (args.length == 0) {
            if (sender instanceof Player player) {
                // Ouvrir le menu principal
                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    new loyfael.gui.ShopMenuGUI(plugin).openMainMenu(player);
                });
                return true;
            } else {
                showHelp(sender);
                return true;
            }
        }

        String subCommand = args[0].toLowerCase();

        switch (subCommand) {
            case "help", "aide" -> {
                showHelp(sender);
                return true;
            }

            case "info" -> {
                if (sender instanceof Player player) {
                    showPlayerInfo(player);
                } else {
                    sender.sendMessage("§cCette commande ne peut être utilisée que par un joueur.");
                }
                return true;
            }

            case "reload" -> {
                if (!sender.hasPermission("aetherplayershop.admin")) {
                    sender.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.");
                    return true;
                }

                plugin.getConfigManager().reload();
                plugin.getCacheManager().invalidateCache();
                sender.sendMessage("§aConfiguration rechargée avec succès!");
                return true;
            }

            case "stats" -> {
                if (!sender.hasPermission("aetherplayershop.admin")) {
                    sender.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.");
                    return true;
                }

                showStats(sender);
                return true;
            }

            case "delete" -> {
                if (!(sender instanceof Player player)) {
                    sender.sendMessage("§cCette commande ne peut être utilisée que par un joueur.");
                    return true;
                }

                if (!player.hasPermission("aetherplayershop.manage")) {
                    player.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.");
                    return true;
                }

                handleDeleteCommand(player, args);
                return true;
            }

            default -> {
                sender.sendMessage("§cCommande inconnue. Utilisez §e/chestshop help §cpour l'aide.");
                return true;
            }
        }
    }

    private void showHelp(@NotNull CommandSender sender) {
        sender.sendMessage("§6§l=== AetherPlayerShop - Aide ===");
        sender.sendMessage("§e/chestshop §7- Ouvre le menu principal des shops");
        sender.sendMessage("§e/chestshop info §7- Affiche vos informations de shops");
        sender.sendMessage("§e/chestshop help §7- Affiche cette aide");

        if (sender.hasPermission("aetherplayershop.manage")) {
            sender.sendMessage("§e/chestshop delete §7- Supprime un de vos shops");
        }

        if (sender.hasPermission("aetherplayershop.admin")) {
            sender.sendMessage("§c/chestshop reload §7- Recharge la configuration");
            sender.sendMessage("§c/chestshop stats §7- Affiche les statistiques du plugin");
        }

        sender.sendMessage("");
        sender.sendMessage("§7Pour créer un shop:");
        sender.sendMessage("§71. Placez un coffre avec les items à vendre");
        sender.sendMessage("§72. Placez une pancarte à côté du coffre");
        sender.sendMessage("§73. Écrivez sur la pancarte:");
        sender.sendMessage("§7   Ligne 1: §e[SELL] §7ou §e[BUY]");
        sender.sendMessage("§7   Ligne 2: §eprix (ex: 10◎)");
        sender.sendMessage("§7   Ligne 3: §equantité (ex: 64)");
        sender.sendMessage("§7   Ligne 4: §elaissez vide");
    }

    private void showPlayerInfo(@NotNull Player player) {
        plugin.getShopManager().getPlayerShops(player.getUniqueId())
            .thenAccept(shops -> {
                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    player.sendMessage("§6§l=== Vos Shops ===");
                    player.sendMessage("§7Nombre de shops: §e" + shops.size() + "§7/§e" +
                        plugin.getConfigManager().getMaxShopsPerPlayer());

                    if (shops.isEmpty()) {
                        player.sendMessage("§7Vous n'avez aucun shop actif.");
                        player.sendMessage("§7Créez un shop en plaçant une pancarte à côté d'un coffre!");
                        return;
                    }

                    player.sendMessage("§7Liste de vos shops:");
                    for (int i = 0; i < Math.min(shops.size(), 10); i++) {
                        var shop = shops.get(i);
                        String status = shop.isActive() ? "§aActif" : "§cInactif";
                        player.sendMessage("§8" + (i + 1) + ". §e" + shop.getItem().getDisplayName() +
                            " §7- " + status + " §7- §e" + String.format("%.2f", shop.getPrice()) + "§6◎");
                    }

                    if (shops.size() > 10) {
                        player.sendMessage("§7... et " + (shops.size() - 10) + " autres shops.");
                    }
                });
            });
    }

    private void showStats(@NotNull CommandSender sender) {
        // Récupération des statistiques du cache
        var cacheStats = plugin.getCacheManager().getCacheStats();

        sender.sendMessage("§6§l=== Statistiques AetherPlayerShop ===");
        sender.sendMessage("§7Cache:");
        sender.sendMessage("§7  - Shops en cache: §e" + cacheStats.totalCached);
        sender.sendMessage("§7  - Entrées expirées: §e" + cacheStats.expiredEntries);
        sender.sendMessage("§7  - Total accès: §e" + cacheStats.totalAccesses);

        // Statistiques MongoDB (asynchrone)
        plugin.getMongoManager().executeAsync(() -> {
            try {
                long totalShops = plugin.getMongoManager().getShopsCollection().countDocuments();
                long activeShops = plugin.getMongoManager().getShopsCollection()
                    .countDocuments(new org.bson.Document("active", true));
                long totalTransactions = plugin.getMongoManager().getTransactionsCollection().countDocuments();

                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    sender.sendMessage("§7Base de données:");
                    sender.sendMessage("§7  - Total shops: §e" + totalShops);
                    sender.sendMessage("§7  - Shops actifs: §e" + activeShops);
                    sender.sendMessage("§7  - Total transactions: §e" + totalTransactions);
                });
            } catch (Exception e) {
                sender.sendMessage("§cErreur lors de la récupération des statistiques MongoDB.");
            }
            return null;
        });
    }

    private void handleDeleteCommand(@NotNull Player player, @NotNull String[] args) {
        if (args.length < 2) {
            player.sendMessage("§cUtilisation: /chestshop delete <ID>");
            player.sendMessage("§7Utilisez §e/chestshop info §7pour voir vos shops avec leurs IDs.");
            return;
        }

        String shopId = args[1];

        // Vérification que le shop appartient au joueur
        plugin.getShopManager().getShop(shopId)
            .thenAccept(shop -> {
                if (shop == null) {
                    player.sendMessage("§cShop introuvable.");
                    return;
                }

                if (!shop.getOwnerUUID().equals(player.getUniqueId()) &&
                    !player.hasPermission("aetherplayershop.admin")) {
                    player.sendMessage("§cCe shop ne vous appartient pas.");
                    return;
                }

                // Suppression du shop
                plugin.getShopManager().deleteShop(shopId)
                    .thenAccept(success -> {
                        plugin.getServer().getScheduler().runTask(plugin, () -> {
                            if (success) {
                                player.sendMessage("§aShop supprimé avec succès!");
                            } else {
                                player.sendMessage("§cErreur lors de la suppression du shop.");
                            }
                        });
                    });
            });
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command,
                                              @NotNull String alias, @NotNull String[] args) {
        if (args.length == 1) {
            List<String> suggestions = Arrays.asList("help", "info");

            if (sender.hasPermission("aetherplayershop.manage")) {
                suggestions.add("delete");
            }

            if (sender.hasPermission("aetherplayershop.admin")) {
                suggestions.add("reload");
                suggestions.add("stats");
            }

            return suggestions.stream()
                .filter(s -> s.toLowerCase().startsWith(args[0].toLowerCase()))
                .collect(Collectors.toList());
        }

        return null;
    }
}

File: AetherPlayerShop2/src/main/kotlin/loyfael/commands/ChestShopCommand.kt
package loyfael.commands

import loyfael.Main
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player
import org.bukkit.permissions.PermissionAttachmentInfo

/**
 * Commande principale du plugin AetherPlayerShop
 */
class ChestShopCommand(private val plugin: Main) : CommandExecutor, TabCompleter {

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<out String>
    ): Boolean {
        if (args.isEmpty()) {
            if (sender is Player) {
                // Ouvrir le menu principal
                plugin.server.scheduler.runTask(plugin, Runnable {
                    plugin.shopMenuGUI.openMainMenu(sender)
                })
                return true
            } else {
                showHelp(sender)
                return true
            }
        }

        val subCommand = args[0].lowercase()

        return when (subCommand) {
            "help", "aide" -> {
                showHelp(sender)
                true
            }

            "info" -> {
                if (sender is Player) {
                    showPlayerInfo(sender)
                } else {
                    sender.sendMessage("§cCette commande ne peut être utilisée que par un joueur.")
                }
                true
            }

            "reload" -> {
                if (!sender.hasPermission("aetherplayershop.admin")) {
                    sender.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.")
                    return true
                }

                plugin.configManager.reload()
                plugin.cacheManager.invalidateCache()
                sender.sendMessage("§aConfiguration rechargée avec succès!")
                true
            }

            "stats" -> {
                if (!sender.hasPermission("aetherplayershop.admin")) {
                    sender.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.")
                    return true
                }

                showStats(sender)
                true
            }

            "delete" -> {
                if (sender !is Player) {
                    sender.sendMessage("§cCette commande ne peut être utilisée que par un joueur.")
                    return true
                }

                if (!sender.hasPermission("aetherplayershop.manage")) {
                    sender.sendMessage("§cVous n'avez pas la permission d'utiliser cette commande.")
                    return true
                }

                handleDeleteCommand(sender, args)
                true
            }

            else -> {
                sender.sendMessage("§cCommande inconnue. Utilisez §e/chestshop help §cpour l'aide.")
                true
            }
        }
    }

    private fun showHelp(sender: CommandSender) {
        sender.sendMessage("§6§l=== AetherPlayerShop - Aide ===")
        sender.sendMessage("§e/chestshop §7- Ouvre le menu principal des shops")
        sender.sendMessage("§e/chestshop info §7- Affiche vos informations de shops")
        sender.sendMessage("§e/chestshop help §7- Affiche cette aide")

        if (sender.hasPermission("aetherplayershop.manage")) {
            sender.sendMessage("§e/chestshop delete §7- Supprime un de vos shops")
        }

        if (sender.hasPermission("aetherplayershop.admin")) {
            sender.sendMessage("§c/chestshop reload §7- Recharge la configuration")
            sender.sendMessage("§c/chestshop stats §7- Affiche les statistiques du plugin")
        }

        sender.sendMessage("")
        sender.sendMessage("§7Pour créer un shop:")
        sender.sendMessage("§71. Remplis un coffre avec §eun seul type d'item")
        sender.sendMessage("§72. Prends une §epancarte §7en main")
        sender.sendMessage("§73. §eClic droit §7sur le coffre (debout) pour ouvrir la création")
        sender.sendMessage("§74. §8Astuce: accroupi = placement normal du panneau")
    }

    private fun showPlayerInfo(player: Player) {
        plugin.shopManager.getPlayerShops(player.uniqueId)
            .thenAccept { shops ->
                plugin.server.scheduler.runTask(plugin, Runnable {
                    val max = getMaxShopsFromPermissions(player, plugin.configManager.maxShopsPerPlayer)
                    player.sendMessage("§6§l=== Vos Shops ===")
                    player.sendMessage("§7Nombre de shops: §e${shops.size}§7/§e$max")

                    if (shops.isEmpty()) {
                        player.sendMessage("§7Vous n'avez aucun shop actif.")
                        player.sendMessage("§7Créez un shop en plaçant une pancarte à côté d'un coffre!")
                        return@Runnable
                    }

                    player.sendMessage("§7Liste de vos shops:")
                    for (i in 0 until minOf(shops.size, 10)) {
                        val shop = shops[i]
                        val status = if (shop.active) "§aActif" else "§cInactif"
                        player.sendMessage(
                            "§8${i + 1}. §e${shop.item.displayName} §7- $status §7- §e${
                                String.format(
                                    "%.2f",
                                    shop.price
                                )
                            }§6◎"
                        )
                    }

                    if (shops.size > 10) {
                        player.sendMessage("§7... et ${shops.size - 10} autres shops.")
                    }
                })
            }
    }

    private fun showStats(sender: CommandSender) {
        // Récupération des statistiques du cache
        val cacheStats = plugin.cacheManager.getCacheStats()

        sender.sendMessage("§6§l=== Statistiques AetherPlayerShop ===")
        sender.sendMessage("§7Cache:")
        sender.sendMessage("§7  - Shops en cache: §e${cacheStats.totalCached}")
        sender.sendMessage("§7  - Entrées expirées: §e${cacheStats.expiredEntries}")
        sender.sendMessage("§7  - Total accès: §e${cacheStats.totalAccesses}")

        // Statistiques MySQL (asynchrone)
        plugin.mySqlManager.supplyAsync<Unit> {
            var totalShops = 0L
            var activeShops = 0L
            var totalTransactions = 0L

            try {
                plugin.mySqlManager.getConnection().use { conn ->
                    conn.prepareStatement("SELECT COUNT(*) FROM shops").use { ps ->
                        ps.executeQuery().use { rs ->
                            if (rs.next()) totalShops = rs.getLong(1)
                        }
                    }
                    conn.prepareStatement("SELECT COUNT(*) FROM shops WHERE active=1").use { ps ->
                        ps.executeQuery().use { rs ->
                            if (rs.next()) activeShops = rs.getLong(1)
                        }
                    }
                    conn.prepareStatement("SELECT COUNT(*) FROM transactions").use { ps ->
                        ps.executeQuery().use { rs ->
                            if (rs.next()) totalTransactions = rs.getLong(1)
                        }
                    }
                }
            } catch (e: Exception) {
                plugin.server.scheduler.runTask(plugin, Runnable {
                    sender.sendMessage("§cErreur lors de la récupération des statistiques MySQL.")
                })
                return@supplyAsync
            }

            val fTotalShops = totalShops
            val fActiveShops = activeShops
            val fTotalTransactions = totalTransactions

            plugin.server.scheduler.runTask(plugin, Runnable {
                sender.sendMessage("§7Base de données:")
                sender.sendMessage("§7  - Total shops: §e$fTotalShops")
                sender.sendMessage("§7  - Shops actifs: §e$fActiveShops")
                sender.sendMessage("§7  - Total transactions: §e$fTotalTransactions")
            })
        }
    }

    private fun handleDeleteCommand(player: Player, args: Array<out String>) {
        if (args.size < 2) {
            player.sendMessage("§cUtilisation: /chestshop delete <ID>")
            player.sendMessage("§7Utilisez §e/chestshop info §7pour voir vos shops avec leurs IDs.")
            return
        }

        val shopId = args[1]

        // Vérification que le shop appartient au joueur
        plugin.shopManager.getShop(shopId)
            .thenAccept { shop ->
                if (shop == null) {
                    player.sendMessage("§cShop introuvable.")
                    return@thenAccept
                }

                if (shop.ownerUUID != player.uniqueId && !player.hasPermission("aetherplayershop.admin")) {
                    player.sendMessage("§cCe shop ne vous appartient pas.")
                    return@thenAccept
                }

                // Suppression du shop
                plugin.shopManager.deleteShop(shopId)
                    .thenAccept { success ->
                        plugin.server.scheduler.runTask(plugin, Runnable {
                            if (success) {
                                player.sendMessage("§aShop supprimé avec succès!")
                            } else {
                                player.sendMessage("§cErreur lors de la suppression du shop.")
                            }
                        })
                    }
            }
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        alias: String,
        args: Array<out String>
    ): List<String>? {
        if (args.size == 1) {
            val suggestions = mutableListOf("help", "info")

            if (sender.hasPermission("aetherplayershop.manage")) {
                suggestions.add("delete")
            }

            if (sender.hasPermission("aetherplayershop.admin")) {
                suggestions.add("reload")
                suggestions.add("stats")
            }

            return suggestions.filter { it.lowercase().startsWith(args[0].lowercase()) }
        }

        return null
    }

    private fun getMaxShopsFromPermissions(player: Player, fallback: Int): Int {
        var max = fallback
        for (pai: PermissionAttachmentInfo in player.effectivePermissions) {
            val perm = pai.permission
            if (!perm.startsWith("aetherplayershop.shops.")) continue
            val tail = perm.substring("aetherplayershop.shops.".length)
            try {
                max = maxOf(max, tail.toInt())
            } catch (ignored: NumberFormatException) {
            }
        }
        return max
    }
}
